ksql-engine/src/main/java/io/confluent/ksql/engine/InsertValuesExecutor.java

```diff
@@ -179,6 +182,10 @@ public void execute(
 
       throw new KsqlException("Failed to insert values into stream/table: "
           + insertValues.getTarget().name(), rootCause);
+    } catch (KsqlException e) {
+      throw new KsqlStatementException(
```

Comment:
`KsqlStatementException ` should only be thrown if there is a problem with the statement itself, not the execution of the statement.  Hence we should think carefully about when we throw one.

Looking at the `execute` method any exception thrown by anything before `producer.send` is a statement error. But the `send` method will only throw if there are issues with producing to kafka, which is a service error.

So, ideally, we'd restructure the code to  throw `KsqlStatementException` on any line before `send`, i.e. we throw on for any issue building the producer record.  Therefore, I'd advise splitting the building of the producer records into its own method, which throws a statement exception, e.g.

```java
public void execute(
      final ConfiguredStatement<InsertValues> statement,
      final KsqlExecutionContext executionContext,
      final ServiceContext serviceContext
  ) {
    final ProducerRecord<byte[], byte[]> record =
        buildRecord(statement, executionContext, serviceContext);

    final InsertValues insertValues = statement.getStatement();

    try {
      final KsqlConfig config = statement.getConfig()
          .cloneWithPropertyOverwrite(statement.getOverrides());

      producer.sendRecord(record, serviceContext, config.getProducerClientConfigProps());
    } catch (final TopicAuthorizationException e) {
      // TopicAuthorizationException does not give much detailed information about why it failed,
      // except which topics are denied. Here we just add the ACL to make the error message
      // consistent with other authorization error messages.
      final Exception rootCause = new KsqlTopicAuthorizationException(
          AclOperation.WRITE,
          e.unauthorizedTopics()
      );

      throw new KsqlException("Failed to insert values into stream/table: "
          + insertValues.getTarget().name(), rootCause);
    } catch (final Exception e) {
      throw new KsqlException("Failed to insert values into stream/table: "
          + insertValues.getTarget().name(), e);
    }
  }

  private ProducerRecord<byte[], byte[]> buildRecord(
      final ConfiguredStatement<InsertValues> statement,
      final KsqlExecutionContext executionContext,
      final ServiceContext serviceContext
  ) {
    final InsertValues insertValues = statement.getStatement();

    try {
      throwIfDisabled(statement.getConfig());

      final KsqlConfig config = statement.getConfig()
          .cloneWithPropertyOverwrite(statement.getOverrides());

      final DataSource<?> dataSource = executionContext
          .getMetaStore()
          .getSource(insertValues.getTarget().name());

      if (dataSource == null) {
        throw new KsqlException("Cannot insert values into an unknown stream/table: "
            + insertValues.getTarget().name());
      }

      if (dataSource.getKsqlTopic().getKeyFormat().isWindowed()) {
        throw new KsqlException("Cannot insert values into windowed stream/table!");
      }

      final RowData row = extractRow(insertValues, dataSource);
      final byte[] key = serializeKey(row.key, dataSource, config, serviceContext);
      final byte[] value = serializeValue(row.value, dataSource, config, serviceContext);

      final String topicName = dataSource.getKafkaTopicName();

      return new ProducerRecord<>(
          topicName,
          null,
          row.ts,
          key,
          value
      );
    } catch (final Exception e) {
      throw new KsqlStatementException(
          "Failed to insert values into " + insertValues.getTarget().name(),
          statement.getStatementText(),
          e
      );
    }
  }
```